File: VB/WinSample.Module/Master.vb
Imports Microsoft.VisualBasic
Imports System

Imports DevExpress.Xpo

Imports DevExpress.ExpressApp
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Validation
Imports DevExpress.Data.Filtering

Namespace WinSample.Module
	<DefaultClassOptions> _
	Public Class Master
		Inherits BaseObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private _MasterName As String
		Public Property MasterName() As String
			Get
				Return _MasterName
			End Get
			Set(ByVal value As String)
				SetPropertyValue("MasterName", _MasterName, value)
			End Set
		End Property
		<Association("Master-Details")> _
		Public ReadOnly Property Details() As XPCollection(Of Detail)
			Get
				Return GetCollection(Of Detail)("Details")
			End Get
		End Property
		Private _SubDetails As XPCollection(Of SubDetail)
		Public ReadOnly Property SubDetails() As XPCollection(Of SubDetail)
			Get
				If _SubDetails Is Nothing Then
					UpdateSubDetails()
				End If
				Return _SubDetails
			End Get
		End Property
		Public Sub UpdateSubDetails()
			If Details.Count > 0 Then
				_SubDetails = New XPCollection(Of SubDetail)(Session, New InOperator("Detail", Details))
			Else
				_SubDetails = Nothing
			End If
		End Sub
		Protected Overrides Overloads Function CreateCollection(Of T)(ByVal [property] As DevExpress.Xpo.Metadata.XPMemberInfo) As XPCollection(Of T)
			Dim collection As XPCollection(Of T) = MyBase.CreateCollection(Of T)([property])
			If [property].Name = "Details" Then
				AddHandler collection.CollectionChanged, AddressOf Details_CollectionChanged
			End If
			Return collection
		End Function
		Private Sub Details_CollectionChanged(ByVal sender As Object, ByVal e As XPCollectionChangedEventArgs)
			UpdateSubDetails()
			OnChanged("SubDetails")
		End Sub
	End Class

End Namespace

File: CS/WinSample.Module/Master.cs
using System;

using DevExpress.Xpo;

using DevExpress.ExpressApp;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Validation;
using DevExpress.Data.Filtering;

namespace WinSample.Module {
    [DefaultClassOptions]
    public class Master : BaseObject {
        public Master(Session session) : base(session) { }
        private string _MasterName;
        public string MasterName {
            get { return _MasterName; }
            set { SetPropertyValue("MasterName", ref _MasterName, value); }
        }
        [Association("Master-Details")]
        public XPCollection<Detail> Details {
            get {
                return GetCollection<Detail>("Details");
            }
        }
        private XPCollection<SubDetail> _SubDetails;
        public XPCollection<SubDetail> SubDetails {
            get {
                if (_SubDetails == null) {
                    UpdateSubDetails();
                }
                return _SubDetails;
            }
        }
        public void UpdateSubDetails() {
            if (Details.Count > 0) {
                _SubDetails = new XPCollection<SubDetail>(Session, new InOperator("Detail", Details));
            } else {
                _SubDetails = null;
            }
        }
        protected override XPCollection<T> CreateCollection<T>(DevExpress.Xpo.Metadata.XPMemberInfo property) {
            XPCollection<T> collection = base.CreateCollection<T>(property);
            if (property.Name == "Details") {
                collection.CollectionChanged += new XPCollectionChangedEventHandler(Details_CollectionChanged);
            }
            return collection;
        }
        void Details_CollectionChanged(object sender, XPCollectionChangedEventArgs e) {
            UpdateSubDetails();
            OnChanged("SubDetails");
        }
    }

}